src/algorithms/crypto/randomness.go

```diff
@@ -1,40 +1,43 @@
 package crypto
 
-import (
-	util "github.com/filecoin-project/specs/util"
-)
+import util "github.com/filecoin-project/specs/util"
+import addr "github.com/filecoin-project/go-address"
+import abi "github.com/filecoin-project/specs/actors/abi"
 
-type Randomness = util.Randomness
+type DomainSeparationTag int
+type Randomness = abi.Randomness
 type Serialization = util.Serialization
 
-// Derive a random byte string from a domain separation tag and an arbitrary
-// serializable object.
-//
-// Note: to produce values of type Serialization, use the auto-generated method prototypes
-//   Serialize_T(T) Serialization
-// for each type T defined in .id.
-//
-// In order to derive randomness from a collection of objects, rather than just a single
-// object, define a struct at the .id level that contains those objects as member fields.
-// This will then cause a Serialize_*() method to be generated for the struct type.
-func DeriveRand(tag DomainSeparationTag, s Serialization) Randomness {
-	return _deriveRandInternal(tag, s, -1)
+const (
+	DomainSeparationTag_TicketDrawing DomainSeparationTag = 1 + iota
+	DomainSeparationTag_TicketProduction
+	DomainSeparationTag_ElectionPoStChallengeSeed
+	DomainSeparationTag_SurprisePoStChallengeSeed
+	DomainSeparationTag_SurprisePoStSelectMiners
+	DomainSeparationTag_SurprisePoStSampleSectors
+)
+
+// Derive a random byte string from a domain separation tag and the appropriate values
+func DeriveRandWithMinerAddr(tag DomainSeparationTag, tix abi.RandomnessSeed, minerAddr addr.Address) Randomness {
+	buffer := _deriveRandInternal(tag, tix, -1)
+	var serializedAddr abi.Bytes
+	util.IMPL_FINISH() // serialize the address
+	buffer = append(buffer, serializedAddr...)
+	ret := SHA256(buffer)
+	return Randomness(ret)
 }
 
-// As in DeriveRand(), but additionally accepts an index into the implicit pseudorandom stream.
-// Index must be strictly positive.
-func DeriveRandWithIndex(tag DomainSeparationTag, s Serialization, index int) Randomness {
-	if index <= 0 {
-		panic("DeriveRandWithIndex only accepts indices > 0")
-	}
-	return _deriveRandInternal(tag, s, index)
+func DeriveRandWithEpoch(tag DomainSeparationTag, tix abi.RandomnessSeed, epoch int) Randomness {
+	buffer := _deriveRandInternal(tag, tix, -1)
+	buffer = append(buffer, BigEndianBytesFromInt(epoch)...)
+	ret := SHA256(buffer)
+	return Randomness(ret)
 }
 
-func _deriveRandInternal(tag DomainSeparationTag, s Serialization, index int) Randomness {
+func _deriveRandInternal(tag DomainSeparationTag, tix abi.RandomnessSeed, index int) util.Bytes {
 	buffer := []byte{}
 	buffer = append(buffer, BigEndianBytesFromInt(int(tag))...)
 	buffer = append(buffer, BigEndianBytesFromInt(int(index))...)
-	buffer = append(buffer, util.Bytes(s)...)
-	ret := SHA256(buffer)
-	return Randomness(ret)
+	buffer = append(buffer, util.Bytes(tix)...)
```

Comment:
[Blocking] We lost the SHA invocation here...

Reply:
Yeah, given the addition of appropriate entropy in the public methods, the SHA invocations have to be done there. It's a drawback of this interface granted.

Reply:
I see, the usage has changed then; in that case, this method definitely shouldn't be called `_deriveRandInternal` as it doesn't derive randomness at all.

(FWIW, the solution I had in mind was to keep `_deriveRandInternal` as is, and have the wrapper functions pass in the serialization of a struct that contains all the relevant parameters. Then the randomness derivation itself is isolated to `_deriveRandInternal`, with the wrappers just packaging up structs.)

Reply:
Changed accordingly! Fair enough